File: frontend/src/lib/instantCrypto.ts
// Instant's ECIES envelope.
//
// ===========================================================================
// INTEROP CONTRACT — the native iOS client must match this byte for byte.
//
//   Device keypair   ECDH P-256. Chosen over X25519 because P-256 is the only
//                    curve the iOS Secure Enclave supports.
//   Content key      AES-256-GCM, random 12-byte IV, 128-bit tag appended to
//                    the ciphertext (WebCrypto's default layout).
//   Key wrapping     One ephemeral P-256 keypair per message. Per recipient
//                    device: ECDH(ephemeral_priv, device_pub) -> 256 raw bits
//                    -> HKDF-SHA256 -> 256-bit wrapping key -> AES-256-GCM
//                    over the raw 32-byte content key, with its own 12-byte IV.
//   HKDF salt        The raw uncompressed ephemeral public key (65 bytes).
//   HKDF info        UTF-8 "eddies-lounge/instant/v1|<senderUserId>|<recipientDeviceId>"
//                    Binding the device id stops an envelope being replayed
//                    against a different device.
//   Encoding         Raw uncompressed P-256 points and all blobs as unpadded
//                    base64url.
//
// CryptoKit equivalent:
//   P256.KeyAgreement.PrivateKey
//     .sharedSecretFromKeyAgreement(with:)
//     .hkdfDerivedSymmetricKey(using: SHA256.self, salt:, sharedInfo:,
//                              outputByteCount: 32)
//   then AES.GCM.seal / AES.GCM.open
// ===========================================================================

import { fromBase64Url, toBase64Url, type InstantDevice } from "./instantKeystore";

export const INSTANT_CRYPTO_VERSION = "eddies-lounge/instant/v1";

const IV_BYTES = 12;
const CONTENT_KEY_BYTES = 32;

export type RecipientDeviceKey = {
  id: number;
  deviceId: string;
  publicKey: string;
};

export type SealedEnvelope = {
  deviceKeyId: number;
  wrappedKey: string;
  wrapIv: string;
};

export type SealedInstant = {
  ciphertext: ArrayBuffer;
  mediaIv: string;
  ephemeralPubKey: string;
  envelopes: SealedEnvelope[];
};

function hkdfInfo(senderUserId: number, recipientDeviceId: string): Uint8Array {
  return new TextEncoder().encode(
    `${INSTANT_CRYPTO_VERSION}|${senderUserId}|${recipientDeviceId}`
  );
}

function randomIv(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(IV_BYTES));
}

async function importDevicePublicKey(publicKeyB64: string): Promise<CryptoKey> {
  return crypto.subtle.importKey(
    "raw",
    fromBase64Url(publicKeyB64) as BufferSource,
    { name: "ECDH", namedCurve: "P-256" },
    true,
    []
  );
}

// ECDH + HKDF. Both sides run this and land on the same 256-bit AES key.
async function deriveWrappingKey(
  privateKey: CryptoKey,
  peerPublicKey: CryptoKey,
  ephemeralPublicKeyRaw: Uint8Array,
  senderUserId: number,
  recipientDeviceId: string
): Promise<CryptoKey> {
  const sharedBits = await crypto.subtle.deriveBits(
    { name: "ECDH", public: peerPublicKey },
    privateKey,
    256
  );
  const hkdfSource = await crypto.subtle.importKey("raw", sharedBits, "HKDF", false, [
    "deriveKey",
  ]);
  return crypto.subtle.deriveKey(
    {
      name: "HKDF",
      hash: "SHA-256",
      salt: ephemeralPublicKeyRaw as BufferSource,
      info: hkdfInfo(senderUserId, recipientDeviceId) as BufferSource,
    },
    hkdfSource,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"]
  );
}

// Encrypts once, then wraps the content key separately for each of the
// recipient's devices. The server sees only ciphertext and wrapped keys.
export async function sealForDevices(
  media: Blob,
  senderUserId: number,
  devices: RecipientDeviceKey[]
): Promise<SealedInstant> {
  if (devices.length === 0) {
    throw new Error("This person has not set up Instant on any device yet.");
  }

  const contentKeyBytes = crypto.getRandomValues(new Uint8Array(CONTENT_KEY_BYTES));
  const contentKey = await crypto.subtle.importKey(
    "raw",
    contentKeyBytes as BufferSource,
    { name: "AES-GCM" },
    false,
    ["encrypt"]
  );

  const mediaIv = randomIv();
  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: mediaIv as BufferSource },
    contentKey,
    await media.arrayBuffer()
  );

  const ephemeral = await crypto.subtle.generateKey(
    { name: "ECDH", namedCurve: "P-256" },
    true,
    ["deriveBits", "deriveKey"]
  );
  const ephemeralPublicRaw = new Uint8Array(
    await crypto.subtle.exportKey("raw", ephemeral.publicKey)
  );

  const envelopes: SealedEnvelope[] = [];
  for (const device of devices) {
    const devicePublicKey = await importDevicePublicKey(device.publicKey);
    const wrappingKey = await deriveWrappingKey(
      ephemeral.privateKey,
      devicePublicKey,
      ephemeralPublicRaw,
      senderUserId,
      device.deviceId
    );
    const wrapIv = randomIv();
    const wrappedKey = await crypto.subtle.encrypt(
      { name: "AES-GCM", iv: wrapIv as BufferSource },
      wrappingKey,
      contentKeyBytes as BufferSource
    );
    envelopes.push({
      deviceKeyId: device.id,
      wrappedKey: toBase64Url(wrappedKey),
      wrapIv: toBase64Url(wrapIv),
    });
  }

  // The plaintext content key never leaves this function.
  contentKeyBytes.fill(0);

  return {
    ciphertext,
    mediaIv: toBase64Url(mediaIv),
    ephemeralPubKey: toBase64Url(ephemeralPublicRaw),
    envelopes,
  };
}

export type OpenableInstant = {
  mediaIv: string;
  ephemeralPubKey: string;
  senderId: number;
  mediaType: string;
  envelope: { wrappedKey: string; wrapIv: string };
};

export async function openInstant(
  ciphertext: ArrayBuffer,
  instant: OpenableInstant,
  device: InstantDevice
): Promise<Blob> {
  const ephemeralPublicRaw = fromBase64Url(instant.ephemeralPubKey);
  const ephemeralPublicKey = await crypto.subtle.importKey(
    "raw",
    ephemeralPublicRaw as BufferSource,
    { name: "ECDH", namedCurve: "P-256" },
    true,
    []
  );

  const wrappingKey = await deriveWrappingKey(
    device.privateKey,
    ephemeralPublicKey,
    ephemeralPublicRaw,
    instant.senderId,
    device.deviceId
  );

  const contentKeyBytes = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: fromBase64Url(instant.envelope.wrapIv) as BufferSource },
    wrappingKey,
    fromBase64Url(instant.envelope.wrappedKey) as BufferSource
  );

  const contentKey = await crypto.subtle.importKey(
    "raw",
    contentKeyBytes,
    { name: "AES-GCM" },
    false,
    ["decrypt"]
  );

  const plaintext = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: fromBase64Url(instant.mediaIv) as BufferSource },
    contentKey,
    ciphertext
  );

  return new Blob([plaintext], { type: instant.mediaType || "image/webp" });
}

// A safety number both people can read aloud to check the server did not
// substitute a key. Order-independent so both sides see the same digits.
export async function safetyNumber(
  myKeys: { publicKey: string }[],
  theirKeys: { publicKey: string }[]
): Promise<string> {
  const material = [...myKeys, ...theirKeys]
    .map((key) => key.publicKey)
    .sort()
    .join("|");
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(material));
  const bytes = new Uint8Array(digest);

  const groups: string[] = [];
  for (let i = 0; i < 12; i += 1) {
    // Five digits per group, from a 16-bit slice of the digest.
    const chunk = ((bytes[i * 2] << 8) | bytes[i * 2 + 1]) % 100000;
    groups.push(chunk.toString().padStart(5, "0"));
  }
  return groups.join(" ");
}

// The fingerprint of just one side, for detecting when someone's keys change.
export async function fingerprintKeys(keys: { publicKey: string }[]): Promise<string> {
  const material = keys
    .map((key) => key.publicKey)
    .sort()
    .join("|");
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(material));
  return toBase64Url(digest);
}

File: ios/Instant/Core/Crypto/InstantCrypto.swift
import CryptoKit
import Foundation

// Instant's ECIES envelope.
//
// ===========================================================================
// INTEROP CONTRACT — this must match frontend/src/lib/instantCrypto.ts byte for
// byte. That file holds the authoritative copy of the comment below.
//
//   Device keypair   ECDH P-256. Chosen over X25519 because P-256 is the only
//                    curve the iOS Secure Enclave supports.
//   Content key      AES-256-GCM, random 12-byte IV, 128-bit tag appended to
//                    the ciphertext (WebCrypto's default layout).
//   Key wrapping     One ephemeral P-256 keypair per message. Per recipient
//                    device: ECDH(ephemeral_priv, device_pub) -> 256 raw bits
//                    -> HKDF-SHA256 -> 256-bit wrapping key -> AES-256-GCM
//                    over the raw 32-byte content key, with its own 12-byte IV.
//   HKDF salt        The raw uncompressed ephemeral public key (65 bytes).
//   HKDF info        UTF-8 "eddies-lounge/instant/v1|<senderUserId>|<recipientDeviceId>"
//   Encoding         Raw uncompressed P-256 points (x963Representation) and all
//                    blobs as unpadded base64url.
//
// Three details are easy to get wrong and fail silently:
//
//   * The HKDF salt is the raw 65 key bytes, NOT the base64url text of them.
//   * WebCrypto emits `ciphertext || tag16` and carries the nonce separately,
//     so `AES.GCM.SealedBox.combined` (which prepends the nonce) is the wrong
//     shape. Split and rejoin explicitly.
//   * `recipientDeviceId` is a lowercase UUID. `UUID().uuidString` is uppercase
//     and would change the info string, producing envelopes nobody can open.
// ===========================================================================

public enum InstantCrypto {
    public static let version = "eddies-lounge/instant/v1"

    static let ivBytes = 12
    static let contentKeyBytes = 32
    static let tagBytes = 16

    // MARK: - Wire types

    /// One recipient device from `GET /api/v1/instant/keys/:userId`.
    public struct RecipientDeviceKey: Sendable, Equatable {
        public let id: Int
        public let deviceId: String
        public let publicKey: String

        public init(id: Int, deviceId: String, publicKey: String) {
            self.id = id
            self.deviceId = deviceId
            self.publicKey = publicKey
        }
    }

    /// One wrapped copy of the content key, addressed to a single device.
    public struct SealedEnvelope: Sendable, Equatable {
        public let deviceKeyId: Int
        public let wrappedKey: String
        public let wrapIv: String

        public init(deviceKeyId: Int, wrappedKey: String, wrapIv: String) {
            self.deviceKeyId = deviceKeyId
            self.wrappedKey = wrappedKey
            self.wrapIv = wrapIv
        }
    }

    public struct SealedInstant: Sendable, Equatable {
        public let ciphertext: Data
        public let mediaIv: String
        public let ephemeralPubKey: String
        public let envelopes: [SealedEnvelope]
    }

    /// Everything `open` needs that is not the ciphertext itself.
    public struct OpenableInstant: Sendable, Equatable {
        public let mediaIv: String
        public let ephemeralPubKey: String
        public let senderId: Int
        public let envelopeWrappedKey: String
        public let envelopeWrapIv: String

        public init(
            mediaIv: String,
            ephemeralPubKey: String,
            senderId: Int,
            envelopeWrappedKey: String,
            envelopeWrapIv: String
        ) {
            self.mediaIv = mediaIv
            self.ephemeralPubKey = ephemeralPubKey
            self.senderId = senderId
            self.envelopeWrappedKey = envelopeWrappedKey
            self.envelopeWrapIv = envelopeWrapIv
        }
    }

    public enum CryptoError: Error, Equatable {
        case noRecipientDevices
        case malformedBase64(String)
        case malformedPublicKey
        case malformedCiphertext
        case decryptionFailed
    }

    // MARK: - Key agreement

    /// The exact bytes WebCrypto passes as HKDF `info`.
    static func hkdfInfo(senderUserId: Int, recipientDeviceId: String) -> Data {
        Data("\(version)|\(senderUserId)|\(recipientDeviceId)".utf8)
    }

    /// ECDH + HKDF. Both sides run this and land on the same 256-bit AES key.
    ///
    /// The sender agrees `ephemeral_priv` with `device_pub`; the recipient
    /// agrees `device_priv` with `ephemeral_pub`. Either way the salt is the
    /// ephemeral public key and the info names the *receiving* device.
    static func deriveWrappingKey(
        agreement: SharedSecret,
        ephemeralPublicKeyRaw: Data,
        senderUserId: Int,
        recipientDeviceId: String
    ) -> SymmetricKey {
        agreement.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: ephemeralPublicKeyRaw,
            sharedInfo: hkdfInfo(senderUserId: senderUserId, recipientDeviceId: recipientDeviceId),
            outputByteCount: 32
        )
    }

    // MARK: - AES-GCM in WebCrypto's layout

    /// Encrypts to `ciphertext || tag`, with the nonce returned separately —
    /// which is what WebCrypto's `crypto.subtle.encrypt` produces.
    static func sealDetached(_ plaintext: Data, key: SymmetricKey, nonce: Data) throws -> Data {
        let box = try AES.GCM.seal(
            plaintext,
            using: key,
            nonce: try AES.GCM.Nonce(data: nonce)
        )
        return box.ciphertext + box.tag
    }

    /// Inverse of `sealDetached`: splits the trailing 16-byte tag back off.
    static func openDetached(_ combined: Data, key: SymmetricKey, nonce: Data) throws -> Data {
        guard combined.count > tagBytes else { throw CryptoError.malformedCiphertext }
        let splitAt = combined.count - tagBytes
        let box = try AES.GCM.SealedBox(
            nonce: try AES.GCM.Nonce(data: nonce),
            ciphertext: combined.prefix(splitAt),
            tag: combined.suffix(tagBytes)
        )
        return try AES.GCM.open(box, using: key)
    }

    static func randomIv() -> Data {
        var bytes = Data(count: ivBytes)
        bytes.withUnsafeMutableBytes { raw in
            _ = SecRandomCopyBytes(kSecRandomDefault, ivBytes, raw.baseAddress!)
        }
        return bytes
    }

    static func decodeRequired(_ value: String, label: String) throws -> Data {
        guard let data = Base64URL.decode(value) else {
            throw CryptoError.malformedBase64(label)
        }
        return data
    }

    // MARK: - Seal

    /// Encrypts the media once, then wraps the content key once per device.
    public static func seal(
        media: Data,
        senderUserId: Int,
        devices: [RecipientDeviceKey]
    ) throws -> SealedInstant {
        guard !devices.isEmpty else { throw CryptoError.noRecipientDevices }

        let contentKey = SymmetricKey(size: .bits256)
        let contentKeyBytes = contentKey.withUnsafeBytes { Data($0) }

        let mediaIv = randomIv()
        let ciphertext = try sealDetached(media, key: contentKey, nonce: mediaIv)

        let ephemeral = P256.KeyAgreement.PrivateKey()
        let ephemeralPublicRaw = ephemeral.publicKey.x963Representation

        var envelopes: [SealedEnvelope] = []
        envelopes.reserveCapacity(devices.count)

        for device in devices {
            let publicKeyData = try decodeRequired(device.publicKey, label: "publicKey")
            guard let devicePublicKey = try? P256.KeyAgreement.PublicKey(
                x963Representation: publicKeyData
            ) else {
                throw CryptoError.malformedPublicKey
            }

            let shared = try ephemeral.sharedSecretFromKeyAgreement(with: devicePublicKey)
            let wrappingKey = deriveWrappingKey(
                agreement: shared,
                ephemeralPublicKeyRaw: ephemeralPublicRaw,
                senderUserId: senderUserId,
                recipientDeviceId: device.deviceId
            )

            let wrapIv = randomIv()
            let wrapped = try sealDetached(contentKeyBytes, key: wrappingKey, nonce: wrapIv)
            envelopes.append(
                SealedEnvelope(
                    deviceKeyId: device.id,
                    wrappedKey: Base64URL.encode(wrapped),
                    wrapIv: Base64URL.encode(wrapIv)
                )
            )
        }

        return SealedInstant(
            ciphertext: ciphertext,
            mediaIv: Base64URL.encode(mediaIv),
            ephemeralPubKey: Base64URL.encode(ephemeralPublicRaw),
            envelopes: envelopes
        )
    }

    // MARK: - Open

    /// Unwraps this device's envelope and decrypts the media.
    public static func open(
        ciphertext: Data,
        instant: OpenableInstant,
        device: DeviceIdentity
    ) throws -> Data {
        let ephemeralPublicRaw = try decodeRequired(instant.ephemeralPubKey, label: "ephemeralPubKey")
        guard let ephemeralPublicKey = try? P256.KeyAgreement.PublicKey(
            x963Representation: ephemeralPublicRaw
        ) else {
            throw CryptoError.malformedPublicKey
        }

        let shared = try device.sharedSecret(with: ephemeralPublicKey)
        let wrappingKey = deriveWrappingKey(
            agreement: shared,
            ephemeralPublicKeyRaw: ephemeralPublicRaw,
            senderUserId: instant.senderId,
            recipientDeviceId: device.deviceId
        )

        let wrapIv = try decodeRequired(instant.envelopeWrapIv, label: "wrapIv")
        let wrappedKey = try decodeRequired(instant.envelopeWrappedKey, label: "wrappedKey")

        let contentKeyBytes: Data
        do {
            contentKeyBytes = try openDetached(wrappedKey, key: wrappingKey, nonce: wrapIv)
        } catch {
            throw CryptoError.decryptionFailed
        }

        let mediaIv = try decodeRequired(instant.mediaIv, label: "mediaIv")
        do {
            return try openDetached(
                ciphertext,
                key: SymmetricKey(data: contentKeyBytes),
                nonce: mediaIv
            )
        } catch {
            throw CryptoError.decryptionFailed
        }
    }
}
